subsys/bluetooth/mesh/rpl.c

```diff
@@ -160,12 +155,23 @@ bool bt_mesh_rpl_check(struct bt_mesh_net_rx *rx,
 	return true;
 }
 
+static void rpl_clear_pending_handler(struct k_work *work)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(replay_list); i++) {
+		clear_rpl(&replay_list[i]);
+	}
+}
+
+static K_WORK_DEFINE(rpl_clear_pending, rpl_clear_pending_handler);
+
 void bt_mesh_rpl_clear(void)
 {
 	BT_DBG("");
 
 	if (IS_ENABLED(CONFIG_BT_SETTINGS)) {
-		schedule_rpl_clear();
+		k_work_submit(&rpl_clear_pending);
```

Comment:
I'm looking for, there are no such race conditions

Reply:
BTW: I don't think to change bt mesh rpl store pending, due this is public api.

Reply:
Also, the Bluetooth thread in zephyr is cooperative, there are don't preemptive.

Reply:
It will. Just steps:
1. bt_mesh_rpl_update\check ---- > schedule_rpl_store (scheduled storage action) ----> bt_mesh_settings_store_schedule
2. IV recovery happens here ----> bt_mesh_rpl_clear (it clears everything immediately now)
3. store_pending  ----> bt_mesh_rpl_pending_store(BT_MESH_ADDR_ALL_NODES) ---> bt_mesh_rpl_pending_store
4. store_pending_rpl (for each entry despite they are empty)

You added the second decision maker about doing something with settings. Since there are 2 decision makers(scheduled task in mesh settings and rpl cleaning) you have to apply kind of synchronization between them to not compete for settings API. 

To avoid this it requires to clean up scheduled rpl task: bt_mesh_settings_store_cancel(BT_MESH_SETTINGS_RPL_PENDING) when you clear rpl
or even do not do anything with settings at all at relay on existing scheduling. To fix issue with RPL during IV recovery you need only to clean up ram representation of RPL. But up yo you how to solve this..

Reply:
Done